src/query/ee/src/storages/fuse/operations/ngram_index.rs

```diff
@@ -0,0 +1,315 @@
+//
+// Licensed under the Elastic License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.elastic.co/licensing/elastic-license
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::BTreeMap;
+use std::collections::HashMap;
+use std::collections::VecDeque;
+use std::sync::Arc;
+
+use databend_common_catalog::plan::Projection;
+use databend_common_catalog::table::Table;
+use databend_common_catalog::table_context::TableContext;
+use databend_common_exception::Result;
+use databend_common_expression::BlockMetaInfoDowncast;
+use databend_common_expression::DataBlock;
+use databend_common_pipeline_core::Pipeline;
+use databend_common_pipeline_sources::AsyncSource;
+use databend_common_pipeline_sources::AsyncSourcer;
+use databend_common_pipeline_transforms::AsyncTransform;
+use databend_common_pipeline_transforms::TransformPipelineHelper;
+use databend_common_sql::executor::physical_plans::MutationKind;
+use databend_common_storages_fuse::index::filters::BlockFilter;
+use databend_common_storages_fuse::io::write_data;
+use databend_common_storages_fuse::io::BlockReader;
+use databend_common_storages_fuse::io::BloomIndexState;
+use databend_common_storages_fuse::io::MetaReaders;
+use databend_common_storages_fuse::io::TableMetaLocationGenerator;
+use databend_common_storages_fuse::operations::BlockMetaIndex;
+use databend_common_storages_fuse::operations::CommitSink;
+use databend_common_storages_fuse::operations::MutationGenerator;
+use databend_common_storages_fuse::operations::MutationLogEntry;
+use databend_common_storages_fuse::operations::MutationLogs;
+use databend_common_storages_fuse::operations::TableMutationAggregator;
+use databend_common_storages_fuse::FuseStorageFormat;
+use databend_common_storages_fuse::FuseTable;
+use databend_query::storages::index::NgramArgs;
+use databend_storages_common_cache::LoadParams;
+use databend_storages_common_io::ReadSettings;
+use databend_storages_common_table_meta::meta::BlockMeta;
+use databend_storages_common_table_meta::meta::ExtendedBlockMeta;
+use databend_storages_common_table_meta::meta::Location;
+use databend_storages_common_table_meta::meta::Statistics;
+use databend_storages_common_table_meta::meta::Versioned;
+use opendal::Operator;
+
+pub async fn do_refresh_ngram_index(
+    fuse_table: &FuseTable,
+    ctx: Arc<dyn TableContext>,
+    _index_name: String,
+    segment_locs: Option<Vec<Location>>,
+    pipeline: &mut Pipeline,
+) -> Result<()> {
+    let Some(snapshot) = fuse_table.read_table_snapshot().await? else {
+        // no snapshot
+        return Ok(());
+    };
+
```

Comment:
First check whether the index name exists. If it doesn't exist, we can be return directly and report an error.